src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -0,0 +1,101 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Tuple
+
+from pants.backend.python.target_types import PythonSources
+from pants.base.specs import AddressSpecs, AscendantAddresses
+from pants.core.util_rules.strip_source_roots import (
+    SourceRootStrippedSources,
+    StripSourcesFieldRequest,
+)
+from pants.engine.addresses import Address
+from pants.engine.collection import DeduplicatedCollection
+from pants.engine.rules import rule
+from pants.engine.selectors import Get, MultiGet
+from pants.engine.target import Targets
+from pants.source.source_root import AllSourceRoots
+
+
+@dataclass(frozen=True)
+class PythonModule:
+    module: str
+
+    @classmethod
+    def create_from_stripped_path(cls, path: PurePath) -> "PythonModule":
+        module_name_with_slashes = (
+            path.parent if path.name == "__init__.py" else path.with_suffix("")
+        )
+        return cls(module_name_with_slashes.as_posix().replace("/", "."))
+
+    def address_spec(self, *, source_root: str) -> AscendantAddresses:
+        """The spec for all candidate targets which could feasibly own the module.
+
+        This uses AscendantAddresses because targets can own files in subdirs (e.g. rglobs). We also
+        use the package path, e.g. `helloworld/util/__init__.py`, rather than the module path to
+        ensure that we capture all possible targets. It is okay if this directory does not actually
+        exist.
+        """
+        module_name_with_slashes = self.module.replace(".", "/")
+        return AscendantAddresses(
+            directory=PurePath(source_root, module_name_with_slashes).as_posix()
+        )
+
+    def possible_stripped_paths(self) -> Tuple[PurePath, ...]:
```

Comment:
Consumers seem to generally manipulate these outputs to put them in a set, so consider making this a generator?